File: windows/SakuraMusic/OverlayManager.cs
namespace SakuraMusic;

/// <summary>Keeps one overlay per visible Apple Music window, aligned frame for frame.</summary>
public sealed class OverlayManager
{
    private readonly Settings _settings;
    private readonly Dictionary<IntPtr, OverlayWindow> _overlays = new();
    private List<IntPtr> _lastOrder = new();

    public OverlayManager(Settings settings) => _settings = settings;

    public int VisibleCount => _overlays.Values.Count(o => o.IsShown);

    public void Sync(IReadOnlyList<TrackedWindow> tracked)
    {
        var ids = tracked.Select(t => t.Handle).ToHashSet();

        foreach (var stale in _overlays.Keys.Where(k => !ids.Contains(k)).ToList())
        {
            _overlays[stale].HideOverlay();
            _overlays[stale].Close();
            _overlays.Remove(stale);
        }

        bool orderChanged = false;
        foreach (var window in tracked)
        {
            bool shouldShow = _settings.Enabled && !window.Occluded;
            if (!_overlays.TryGetValue(window.Handle, out var overlay))
            {
                overlay = new OverlayWindow(_settings);
                _overlays[window.Handle] = overlay;
                orderChanged = true;
            }

            overlay.MoveTo(window.Frame);
            overlay.CornerRadius = window.FillsScreen ? 0 : 8;

            if (shouldShow && !overlay.IsShown)
            {
                overlay.ShowOverlay();
                orderChanged = true;
            }
            else if (!shouldShow && overlay.IsShown)
            {
                overlay.HideOverlay();
            }
        }

        var order = tracked.Select(t => t.Handle).ToList();
        if (orderChanged || !order.SequenceEqual(_lastOrder))
        {
            for (int i = order.Count - 1; i >= 0; i--)
            {
                if (_overlays.TryGetValue(order[i], out var o) && o.IsShown) o.BringToTop();
            }
            _lastOrder = order;
        }
    }

    public void ApplySettings()
    {
        foreach (var overlay in _overlays.Values)
        {
            overlay.ApplySettings();
            if (!_settings.Enabled) overlay.HideOverlay();
        }
    }

    public void CloseAll()
    {
        foreach (var overlay in _overlays.Values) overlay.Close();
        _overlays.Clear();
    }
}

File: macos/Sources/SakuraMusic/OverlayManager.swift
import AppKit

/// Keeps one overlay window per visible Music window, aligned frame-for-frame.
final class OverlayManager {
    private let settings: Settings
    private var overlays: [CGWindowID: OverlayWindow] = [:]
    private var lastOrder: [CGWindowID] = []

    init(settings: Settings) {
        self.settings = settings
    }

    var visibleCount: Int {
        overlays.values.filter { $0.isVisible }.count
    }

    func sync(with tracked: [TrackedWindow]) {
        let trackedIDs = Set(tracked.map(\.id))

        // Music window went away (closed, minimised, other Space): drop its overlay.
        for (id, overlay) in overlays where !trackedIDs.contains(id) {
            overlay.orderOut(nil)
            overlays[id] = nil
        }

        var orderChanged = false
        for window in tracked {
            let shouldShow = settings.enabled && (!window.occluded || settings.alwaysShow)
            let overlay: OverlayWindow
            if let existing = overlays[window.id] {
                overlay = existing
            } else {
                overlay = OverlayWindow(frame: window.frame, settings: settings)
                overlays[window.id] = overlay
                orderChanged = true
            }

            if overlay.frame != window.frame {
                overlay.setFrame(window.frame, display: true)
            }
            overlay.overlayView.cornerRadius = window.fillsScreen ? 0 : 10

            if shouldShow, !overlay.isVisible {
                overlay.orderFrontRegardless()
                orderChanged = true
            } else if !shouldShow, overlay.isVisible {
                overlay.orderOut(nil)
            }
        }

        // Mirror Music's own z-order: bring overlays forward back-to-front.
        let order = tracked.map(\.id)
        if orderChanged || order != lastOrder {
            for id in order.reversed() {
                if let overlay = overlays[id], overlay.isVisible {
                    overlay.orderFrontRegardless()
                }
            }
            lastOrder = order
        }
    }

    func applySettings() {
        for overlay in overlays.values {
            overlay.overlayView.applySettings()
            if !settings.enabled { overlay.orderOut(nil) }
        }
    }
}
